Rails「コメントを付けられるようにする」プラクティス

k-nakamura1984 opened the pull request:
reportモデルのCRUDをscaffolで作成しました。

k-nakamura1984 committed:
Rails「コメントを付けられるようにする」プラクティス

reportモデルのCRUDをscaffolで作成しました。

k-nakamura1984 committed:
ReportモデルのCRUDを実装しました。

reportを更新できるのは、そのreportを投稿した本人のみ。
タイムゾーンをTokyoに指定し、表示方法を変更しました。

k-nakamura1984 committed:
Reportモデルにcomment機能を実装しました

k-nakamura1984 committed:
ReportモデルとBooksモデルにポリモーフィックでcomment機能を実装

betachelsea commented on db/migrate/20210831091134_add_polymorphic_to_comments.rb:
```diff
@@ -0,0 +1,9 @@
+class AddPolymorphicToComments < ActiveRecord::Migration[6.1]
+  def change
+    remove_index :comments, :report_id
+    remove_column :comments, :report_id, integer
```
[MUST] この書き方は例外が出てしまっているようですね 👀 
また、まだ元のブランチにmergeされていない開発環境のmigrationファイルについては、編集しなおすという手段もあるのでこの機会に試してみてはどうかと思いました。
[【Rails】$rails db:rollbackしたい時の間違えない手順 \- Qiita](https://qiita.com/s_tatsuki/items/3e1f119c91e21b8f0c33)
rollbackは本番環境(production, 実際にユーザーが利用しているサーバを指します)では滅多(ほぼ0?)しませんが、開発環境ではこのように「1回migrationファイル作ったけどやっぱり直したい！」と言うときにはよくやります。

今回、新しいmigrationファイルが4つありますが、減らしてみてもらえますでしょうか？
方針としては「日報のmigrationファイル」「コメントのmigrationファイル」の2ファイルにまとめるとスッキリするかなと思いました。

k-nakamura1984 replied:
rails db:rollbackをした後、途中変更に関するファイル２つを削除しました。
元々あった日報テーブルの新規作成、コメント新規作成のmigrationファイルに、項目を追記してファイルをまとめました。

betachelsea commented on app/views/reports/show.html.erb:
```diff
@@ -0,0 +1,28 @@
+  <td><%= link_to t('views.common.edit'), edit_report_path(@report) %></td>
+<% else %>
+  <td></td>
+<% end %>
```
[NITS] `<td>` タグはテーブルセルを表すので、この箇所(tableが無い状態)で出てくるのはHTMLとしては変だなと感じました。また、この位置について空のタグが特に用が無い場合は、else節を省略しても良さそうです〜。

betachelsea commented on app/views/reports/show.html.erb:
```diff
@@ -0,0 +1,28 @@
+
+<p>
+  <strong><%= Report.human_attribute_name(:created_time) %></strong>
+  <%= @report.created_at.to_s(:datetime_jp) %>
```
独自time format! いいですね 👍  ✨

betachelsea commented on app/views/reports/index.html.erb:
```diff
@@ -0,0 +1,30 @@
+  </tbody>
+</table>
+
+<br>
```
[NITS] 改行以外の目的での `<br>` の使用は避けましょう〜 👀   レイアウトが気になる場合はcssで対応がGoodです！

betachelsea commented on app/views/comments/_comments.html.erb:
```diff
@@ -0,0 +1,16 @@
+
+<% commentable.comments.each do |comment| %>
+  <div>
+    <a href="/users/<%= comment.user.id %>">
```
[NITS] リンクはURLヘルパーを使って `user_path(comment.user)` とすると、よりrailsっぽいと思いました〜 👀 
また、複数行を link_to で取りたい時は次のようにできます。

```erb
<%= link_to user_path(comment.user) do %>
  <% if comment.user.name.blank? %>
    <%= comment.user.email %>
  <% else %>
    <%= comment.user.name %>
  <% end %>
<% end %>
```

k-nakamura1984 replied:
複数行をlink_toで取る場合、なぜ`do`を記入する必要があるのか情報を見つけることができませんでした。

例えば、Railsドキュメントですと、どの箇所から上記が必要・可能であることを推測できるのでしょうか。
https://railsdoc.com/page/link_to

ヒントがなければ解けなかっと心配しております…。

betachelsea replied:
個別の機能についてはぜひ「Ruby on Rails API」を積極的に参照していただくと良いかなと思います！今回のケースではこちらにブロックを取る例があります 👍   
https://api.rubyonrails.org/classes/ActionView/Helpers/UrlHelper.html#method-i-link_to

railsdocは有志の方による日本語情報なので読みやすいのですが、やはり詳細を掘っていきたいときは英語公式のリファレンスをあたるような癖をどんどんつけていきましょう〜 💪

betachelsea commented on app/views/comments/_comments.html.erb:
```diff
@@ -0,0 +1,16 @@
+      <% if comment.user.name.blank? %>
+        <%= comment.user.email %>
+      <% else %>
+        <%= comment.user.name %>
```
[NITS] AがblankならBを代入、は次のようにできます 👀 

```ruby
 <%= comment.user.name.presence || comment.user.email %>
```

参考: [\[初心者向け\] RubyやRailsでリファクタリングに使えそうなイディオムとか便利メソッドとか \- Qiita](https://qiita.com/jnchito/items/dedb3b889ab226933ccf#%E7%A9%BA%E3%81%AA%E3%82%89%E5%88%A5%E3%81%AE%E5%80%A4%E3%82%92%E4%BB%A3%E5%85%A5%E3%81%AE%E4%BB%A3%E3%82%8F%E3%82%8A%E3%81%ABpresence%E3%82%92%E4%BD%BF%E3%81%86)

betachelsea commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,19 @@
+    @comment = @commentable.comments.new(comment_params)
+    @comment.user = current_user
+    if @comment.save
+      redirect_to @commentable, noteice: 'Your comment was succewwfully posted'
```
[MUST] `noteice` は `notice` にすると、うまくflashが表示されそうです。また、`successfully` でしょうか。あとはすでにi18nのプラクティスを完了済みなので、適用できているとよりniceですね！ 
(※i18nは必須要件に無いようなので修正は必ずしも求めませんが、flash & 英文メッセージの typoは修正していただいた方が良いなと思っております。)

betachelsea commented on app/models/user.rb:
```diff
@@ -12,6 +12,9 @@ class User < ApplicationRecord
   has_one_attached :avatar
 
+  has_many :reports, dependent: :destroy
+  has_many :comments, dependent: :destroy
```
[IMO] 頂いていたコメントについて確認しました。

> userモデルをhas_many :comments, dependent: :nullifyとした場合のエラーが解決できなかったため、has_many :comments, dependent: :destroyとしております。


こちらのエラーというのは、コメント済みのユーザを削除しようとしたときの `ActiveRecord::NotNullViolation` でしょうか？

![スクリーンショット 2021-09-12 16 50 50](https://user-images.githubusercontent.com/2603449/132978327-72b5ffb4-7740-45fe-ac76-d3b3ae95e7f4.png)

もしかすると、migrationでDBに `comment.user_id` を `null: false` にしているために発生する問題かなと思いました。もし、 `ユーザーが退会した後も、該当ユーザーのコメントは残っていた方が良い` という仕様を実現したい場合は、DB側もnullを許す形にする必要がありそうです。
また、`comment.user_id` にnullを許すと言うことは、`comment.user.name` の呼び出しもエラーとなってしまう...などの注意点を考慮して作っていく必要が発生します。

現在指定されている `dependent: destroy` が一番シンプルな状態が保てると思いますので、プラクティスとしてはこれで良いと思います。しかし、もちろん実際のプロダクトになってくると「ユーザーが退会した後も、該当ユーザーのコメントは残っていた方が良い」という仕様を実現する必要に迫られることも、もちろんありそうです。nullifyの指定にチャレンジしてみるかどうかはお任せしますが、もしチャレンジする場合は上記の注意点をチェックしながら進めていただければと思います！

k-nakamura1984 replied:
自信がないながら、いただいたヒントを元に修正しました。
アカウント削除をした後は、「退会ユーザー」と表示するようにできたのですが、これだけで良いのかちょっと不安です😅

betachelsea replied:
退会ユーザの取り扱いの具体については要件に入っていないので、エラーが出ていなくて、 k-nakamura1984さんがその挙動に納得していればOKだと思います！ 👍

betachelsea commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
+
+class ReportsController < ApplicationController
+  before_action :set_report, only: %i[show edit update destroy]
+  before_action :ensure_user, only: %i[edit update destroy]
```
[NITS] メソッドの中身を見ると、set_report で定義した `@report` を ensure_user では上書きしているので、`%[edit update destroy]` では set_report は呼ばなくても良さそうだと思いました。と、なるとset_reportを呼び出しているのは `show` だけになるので、もしかすると `set_report` メソッドの中身は show に移転して無くしてしまうのもありかもしれません〜 👀

betachelsea reviewed (commented):
いくつかコメントさせていただきました！ [MUST]については修正必須として対応をお願いします！ 🙏

k-nakamura1984 committed:
レビュー箇所を修正

不要なtd,brタグを削除。
リンクをlink_to user_pathで書き換え。
AがblankならBを代入の箇所を書き換え。
comments_controller.rbのtypoを修正。i18nに対応。
reports_controller.rbのset_reportメソッドをshowに移転。
migrationとdependent: :nullifyの指摘箇所は次のコミットにて対応予定。

k-nakamura1984 committed:
レビュー箇所を修正

途中変更のあったmigrationファイルを集約

k-nakamura1984 committed:
レビュー箇所を修正

アカウント削除した場合も、コメントデータを保持。

k-nakamura1984 committed:
不要なコメントを削除しました

betachelsea commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,54 @@
+    @report = Report.new(report_params)
+    @report.user_id = current_user.id
+    if @report.save
+      redirect_to @report, notice: 'Report was successfully created.'
```
[NITS] 全体をi18nしようと思ったら、この辺りのメッセージの日本語化も対応していきましょう〜 🇯🇵

betachelsea replied:
[NITS] i18nを全体に適用していこうとする場合、こういうところのメッセージにも注意していきましょう〜 🇯🇵

betachelsea commented on app/views/comments/_comments.html.erb:
```diff
@@ -0,0 +1,17 @@
+<% commentable.comments.each do |comment| %>
+  <div>
+    <% if comment.user.present? %>
+      <%= link_to user_path(comment.user) do %>
+        <%= comment.user.name.presence || comment.user.email %>
+      <% end %>
```
[FYI] do~endから脱却してスッキリしたいという場合、Userモデルに name or email を返すようなメソッドを定義して、それを呼び出す形にするのも良さそうですね〜 👀 

```ruby
class User < ApplicationRecord
  # ... 諸々中身は省略...

  def name_or_email
    name.presence || email
  end

end
```

```erb
<%= link_to comment.user.name_or_email, user_path(comment.user) %>
```

betachelsea reviewed (approved):
修正ありがとうございました！確認OKです 🎉 
少しコメント追加しました〜 📝

k-nakamura1984 committed:
i18nの対応漏れを修正
